### PR TITLE
FD/iPDB sampling technique.

### PR DESCRIPTION
**Pull request by haz**
Coverage Badge

### CODE REVIEW

**Comment by haz** [comment]
Closes #6

File: macq/generate/pddl/vanilla_sampling.py [hunk 1]
```diff
@@ -112,6 +112,8 @@ def generate_single_trace(self=self, plan_len=plan_len):
 
             if not plan_len:
                 plan_len = self.plan_len
+            if callable(plan_len):
```
Comment (e-cal) [inline macq/generate/pddl/vanilla_sampling.py:115]: Following from the above comment, if a callable `plan_len` is called right away, why not call it in the initialization of the `FDRandomWalkSampling`?

**Review by e-cal — COMMENTED** [review]
LGTM apart from some confusion around `plan_len` in the new generator. Could be missing the bigger picture on the intended workflow though.

File: macq/generate/pddl/fd_random_walk.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            #   (does nothing on unit-cost problems).
+            #
+            #   The expected walk length is np = 2 * estimated number of solution steps.
+            #   (We multiply by 2 because the heuristic is underestimating.)
+            avg_op_cost = self._avg_op_cost()
+            assert avg_op_cost > 0, "Average operator cost must be greater than 0"
+            sol_steps = int((init_h / avg_op_cost) * 0.5)
+            self.init_h = 4 * sol_steps
+
+        self.plan_len = self._plan_len
```
[the remark this answers is not in the document]
Comment (haz) [reply macq/generate/pddl/fd_random_walk.py:88]: It shouldn't. Idea is for `self.plan_len` to be a function here.

File: macq/generate/pddl/vanilla_sampling.py (hunk 1 above)
Comment (haz) [reply macq/generate/pddl/vanilla_sampling.py:115]: The key thing is that plan length is _repeatedly_ called -- it's sampling the length each time a new trace is called for.

Comment (e-cal) [reply macq/generate/pddl/vanilla_sampling.py:115]: oh I see, `_plan_len` is non-deterministic that's the piece I was missing

**Pull request merged**